src/iam/nodemanager/nodemanager.cpp

```diff
@@ -0,0 +1,227 @@
+Error NodeManager::Init(NodeInfoStorageItf& storage)
+{
+    mStorage = &storage;
+
+    StaticArray<StaticString<cNodeIDLen>, cNodeMaxNum> nodeIds;
+
+    auto err = storage.GetAllNodeIds(nodeIds);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    NodeInfo nodeInfo;
+
+    for (const auto& nodeId : nodeIds) {
+        err = storage.GetNodeInfo(nodeId, nodeInfo);
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        err = mNodeInfoCache.PushBack(nodeInfo);
+        if (!err.IsNone()) {
+            return err;
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error NodeManager::SetNodeInfo(const NodeInfo& info)
+{
+    Error err = ErrorEnum::eNone;
+
+    if (info.mStatus == NodeStatusEnum::eUnprovisioned) {
+        err = mStorage->RemoveNodeInfo(info.mID);
+        if (err.Is(ErrorEnum::eNotFound)) {
+            err = ErrorEnum::eNone;
+        }
+    } else {
+        err = mStorage->SetNodeInfo(info);
+    }
+
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return UpdateCache(info);
+}
+
+Error NodeManager::SetNodeStatus(const String& nodeId, NodeStatus status)
+{
+    Error err = ErrorEnum::eNone;
+
+    if (status == NodeStatusEnum::eUnprovisioned) {
+        err = mStorage->RemoveNodeInfo(nodeId);
+        if (err.Is(ErrorEnum::eNotFound)) {
+            err = ErrorEnum::eNone;
+        }
+    } else {
+        err = mStorage->SetNodeStatus(nodeId, status);
+    }
+
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return UpdateCache(nodeId, status);
+}
+
+Error NodeManager::GetNodeInfo(const String& nodeId, NodeInfo& nodeInfo) const
+{
+    auto cachedInfo = GetNodeFromCache(nodeId);
+
+    if (cachedInfo == nullptr) {
+        return ErrorEnum::eNotFound;
+    }
+
+    nodeInfo = *cachedInfo;
+
+    return ErrorEnum::eNone;
+}
+
+Error NodeManager::GetAllNodeIds(Array<StaticString<cNodeIDLen>>& ids) const
+{
+    for (const auto& nodeInfo : mNodeInfoCache) {
+        auto err = ids.PushBack(nodeInfo.mID);
+        if (!err.IsNone()) {
+            return err;
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error NodeManager::RemoveNodeInfo(const String& nodeId)
+{
+    auto cachedInfo = GetNodeFromCache(nodeId);
+    if (cachedInfo == nullptr) {
+        // Cache contains all the entries, so if not found => just return error.
+        return AOS_ERROR_WRAP(ErrorEnum::eNotFound);
+    }
+
+    auto err = mStorage->RemoveNodeInfo(nodeId);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = mNodeInfoCache.Remove(cachedInfo).mError;
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (mNodeInfoListener) {
+        mNodeInfoListener->OnNodeRemoved(nodeId);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error NodeManager::SubscribeNodeInfoChange(NodeInfoListenerItf& listener)
+{
+    for (const auto& nodeInfo : mNodeInfoCache) {
```

Comment:
It is not expected that client will receive all updates on subscribe. Usually client will do something like that:
```c++
SubscribeNodeInfoChange
GetAllNodeInfos
```

Reply:
Ok, So the question is whether it is required to send initial value when the listener subscribed to the manager.
I think the current approach is simpler for the manager & its clients, & as a result it is more correct. Unless there are some counterexamples from other frameworks, which might mean that there is a general approach & it differs from the implemented. So I suggest to leave it as is.

Reply:
Agreed not to send nodeInfo on subscription => the client should getallnodeinfo before subscription. Because there is no clear understanding when all the nodes are sent